### ISSUE TITLE
☂️ Reinforcement: Refactor widgets from design into the core before decoupling

### ISSUE THREAD

**Issue by filleduchaos** [flutter/flutter#53059]
This issue tracks the essential refactoring work required to strengthen the core widgets library _**before**_ decoupling the material and cupertino libraries (https://github.com/flutter/flutter/issues/101479).

We will prioritize net-new additions **_after_** decoupling. 
**🔍 For feedback on net-new widgets to strengthen the core, see: link**

---

Our goal is to identify and move foundational, style-agnostic logic out of the existing design system libraries and into the core framework. This will ensure the core is robust enough to support multiple high-quality design systems and will provide a richer set of primitives for all developers.

> [!NOTE]
It is important to note that not _everything_ in the design libraries will be refactored into the core. Some widgets are too heavily tied to their design language to extract a meaningful, un-opinionated API that would satisfy developer needs. One of our guiding principles is to avoid the "[lowest common denominator](https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#avoid-the-lowest-common-denominator)" problem, ensuring that any new core widget is genuinely powerful and flexible. 

### Community Feedback Needed 🗣️

This is a critical opportunity for community input. Please use this issue to provide feedback and suggest other widgets or functionalities you believe are fundamental and should be moved into the core widgets library before the main decoupling occurs.

#### Widgets we are investigating/refactoring before decoupling:

- https://github.com/flutter/flutter/issues/172929
- https://github.com/flutter/flutter/issues/172930
- https://github.com/flutter/flutter/issues/172931

#### Widgets that are unlikely to be refactored, will migrate as-is:

- **ColorScheme**
  - Very material, uses material color utilities package to generate material spec color scheme.
- **InkWell**
  - Relies on Material ancestor to paint, too complex to refactor to a basic component that paints in response to a gesture, which could be done in a different/simpler way
- **Scaffold**
  - Very material specific for layout. We discussed focusing on simple app bar, bottom bar widgets might be nice to have instead with built in safe area handling. A column could compose something similar to a Scaffold layout if the simple *bar components handled things like SafeArea out of the box. Simple composition from the widgets layer is still our focus.
- **SliverAppBar**
  - Highly complex and material-specific
  - SliverPersistentHeader, its base class is already in widgets, We added several atomic sliver header widgets last year that are meant to replace the SliverAppBar behemoth through composition in the widgets layer.
    - [SliverResizingHeader](https://api.flutter.dev/flutter/widgets/SliverResizingHeader-class.html): Pins its child and resizes it from its max extent to its min extent when scrolled
    - [SliverFloatingHeader](https://api.flutter.dev/flutter/widgets/SliverFloatingHeader-class.html): Hides the child when the user scrolls backwards and shows its child when the user scrolls forward
    - [PinnedHeaderSliver](https://api.flutter.dev/flutter/widgets/PinnedHeaderSliver-class.html): Keeps its child pinned to the top of the scroll view.
- **TabBar**
  - Does not implement navigation, instead a scrolling PageView under the hood. Could develop something simpler in the widgets layer, but the current material implementation does not appear to have a clear common base. Very specific to material design (similar assumptions in AppBar, SliverAppBar) with regards to preferred size. We should consider the opportunity for a net-new base component in https://github.com/flutter/flutter/issues/97496.
- **Theme, TextTheme**
  - We discussed a new base theme object as making sense. Added to net-new issue for consideration later (https://github.com/flutter/flutter/issues/97496). For use in the WidgetsApp, containing:
    - ThemeExtensions, or something similar
    - High contrast (a11y)
    - Brightness (dark mode)
    - defaultTargetPlatform 
    - Font family and fallback
    - Perhaps others…

#### To consider - provide feedback and we’ll add to this list:

- **Tooltip**
- **Indicators - progress, linear, refresh**
- **Date pickers - parts of**

---

<details> <summary> OP </summary>

## Use case

The number of platforms that Flutter officially targets to a reasonable level of stability has grown beyond just Android and iOS to include the web, Windows, macOS, and Linux. Yet most of the widgets the SDK provides are still tied to the design guidelines/systems of those two mobile OSes. This seriously impacts the out-of-box usability of the widget library for developers who are using Flutter for other platforms/who do not necessarily prioritise mobile.

I think there is a strong case to be made that there are a lot of widgets in `flutter/material` and `flutter/cupertino` that really should be/be based on flexible, extensible widgets in `flutter/widgets` so that developers can take direct advantage of logic that's often hidden away in private classes and hardcoded constants, rather than having to duplicate it.

## Proposal

Much like with the new plugin system, I think it would be a great idea if certain Material/Cupertino widgets were rewritten to decouple the general "behaviour" of the widget from design-specific guidelines (and probably make more use of delegates and painters). A prime candidate would be `Slider` and `CupertinoSlider`, for example.

One set of widgets that I think already does this decently well is the set of `Scrollbar` (from `flutter/material`), `CupertinoScrollbar` (from `flutter/cupertino`) and `ScrollbarPainter` (from `flutter/widgets`). With the latter it is easy to build a custom scrollbar without having to think too much about (or copy the logic of) how a scrollbar should work, _and_ you also have built-in scrollbars that follow Material guidelines or HIG if that's what you want. The only thing I would change about this set is to update the documentation of `ScrollbarPainter` with a minimal example of how to put together a scrollbar with it.

</details>

**Comment by HansMuller** [comment]
We've often talked about creating a "building blocks" library in between widgets and material/cupertino/etc, to make it easier to build things like the components in the material/cupertino/etc libraries.  That said, at the moment there's no such project in the plan, but we have been thinking about it :-).

That said, appreciate the feedback about `ScrollbarPainter`; an example sounds like a good idea.

CC @justinmc

**Comment by justinmc** [comment]
I definitely agree with this!  I think this is going to become more and more important for Flutter over time.  I will try to update this issue if I see any concrete steps happening anywhere.

**Comment by lukepighetti** [comment]
`material` and `cupertino` should be merely pub packages and flutter should expose a mid-level API for widgets and especially slivers

**Comment by justinmc** [comment]
That was one of my first thoughts when I started working with Flutter.  There might be more support for that kind of move as we add more platform design styles.  A big positive of keeping them in the library now is that it's so easy to write a Flutter app with everything you need right out of the box.  Maybe we'd have to include the material library by default but make it really easy and obvious to swap in/out Cupertino or others.

**Comment by lukepighetti** [comment]
@justinmc They can just add "material" and "Cupertino" to pubspec.yaml dependencies

**Comment by satvikpendem** [comment]
I also agree, a lot of values are hard-coded into certain widgets without a way to overwrite them, such as the box shadow with `ReorderableListView`, which shouldn't be a Material widget but more a generalized list widget that supports different theming.

**Comment by darrenaustin** [comment]
Hey folks, we are ramping up a design to try to address this and other issues with `ReorderableListView`. We are tracking it with #66080 and in a [design doc](https://docs.google.com/document/d/1JzNtMQ-jnPnSHEBoi6IO0x2qMMylwhvRslEDcmqiwSs/edit#heading=h.pub7jnop54q0). Feedback welcome.

**Comment by Hixie** [comment]
I've been noodling on this topic recently and wrote some thoughts up here: https://flutter.dev/go/blankcanvas
Comments welcome.

**Comment by mateusfccp** [comment]
@Hixie I like very much the idea proposed in the mentioned document.

Is there any planning about when will this be done? Or is it still just in discussion phase?

**Comment by Hixie** [comment]
I've been working on a proof of concept (I have buttons and scrollbars implemented so far) but it's more of a hobby than something I'm working on full time currently. When there's something to report I'll leave a note at the top of the design doc.

**Comment by iapicca** [comment]
this issue has been **indirectly** referenced in [2024 road map](https://github.com/flutter/flutter/wiki/Roadmap) as `related issue` of  [blankcanvas](https://flutter.dev/go/blankcanvas)

**Comment by Piinks** [comment]
Hi folks! We have decided to pursue decoupling the design libraries from the framework. Please see OP above for how we are coordinating work _before_ decoupling, and linked issues for tracking various aspects of this project.
Parent tracking issue: https://github.com/flutter/flutter/issues/101479

**Comment by mglsj** [comment]
How about moving the (Material) [Icons](https://api.flutter.dev/flutter/material/Icons-class.html) out to a separate package like `cupertino_icons`

**Comment by Piinks** [comment]
Closing this, as this specific period of decoupling (refactoring while all the code is together before moving the libraries to the new packages) is now ended.

**Comment by Piinks** [comment]
> > Our goal is to identify and move foundational, style-agnostic logic out of the existing design system libraries and into the core framework. This will ensure the core is robust enough to support multiple high-quality design systems and will provide a richer set of primitives for all developers.
> 
> Hi [@Piinks[<img alt="" width="15" height="15" src="chrome-extension://dlebflppeeemcdpidccbiblndppbmjmh/ospo-chrome-ext-logo.png">](https://teams.googleplex.com/katelovett@google.com)](https://github.com/Piinks)! Is there an overall list of what was moved to the widgets layer as a core widget/primitive?

Good question! I don't know that there is a final list yet, we'll continue this work by adding new core widgets to the framework going forward and roll out a full catalog. This primarily focused on what we wanted to keep before moving the libraries out, most listed above in the OP or visible as linked PRs. See https://github.com/flutter/flutter/issues/97496 for continued tracking on core widgets.
